sd/etcdv3/client.go

```
@@ -0,0 +1,234 @@
+	resp, err := c.kv.Get(c.ctx, key, clientv3.WithPrefix())
+	if err != nil {
+		return nil, err
+	}
+
+	entries := make([]string, len(resp.Kvs))
+	for i, kv := range resp.Kvs {
+		entries[i] = string(kv.Value)
+	}
+
+	return entries, nil
+}
+
+// WatchPrefix implements the etcd Client interface.
+func (c *client) WatchPrefix(prefix string, ch chan struct{}) {
+	c.wctx, c.wcf = context.WithCancel(c.ctx)
+	c.watcher = clientv3.NewWatcher(c.cli)
+
+	wch := c.watcher.Watch(c.wctx, prefix, clientv3.WithPrefix(), clientv3.WithRev(0))
+	ch <- struct{}{}
+	for wr := range wch {
+		if wr.Canceled {
+			return
+		}
+		ch <- struct{}{}
+	}
+}
+
+func (c *client) Register(s Service) error {
+	var err error
+
+	if s.Key == "" {
+		return ErrNoKey
+	}
+	if s.Value == "" {
+		return ErrNoValue
+	}
+
+	if c.leaser != nil {
+		c.leaser.Close()
+	}
+	c.leaser = clientv3.NewLease(c.cli)
+
+	if c.watcher != nil {
+		c.watcher.Close()
+	}
+	c.watcher = clientv3.NewWatcher(c.cli)
+	if c.kv == nil {
+		c.kv = clientv3.NewKV(c.cli)
+	}
+
+	if s.TTL == nil {
+		s.TTL = NewTTLOption(time.Second*3, time.Second*10)
+	}
+
+	grantResp, err := c.leaser.Grant(c.ctx, int64(s.TTL.ttl.Seconds()))
+	if err != nil {
+		return err
+	}
+	c.leaseID = grantResp.ID
+
+	_, err = c.kv.Put(
+		c.ctx,
+		s.Key,
+		s.Value,
+		clientv3.WithLease(c.leaseID),
+	)
+	if err != nil {
+		return err
+	}
+
+	// this will keep the key alive 'forever' or until we revoke it or
+	// the context is canceled
+	c.hbch, err = c.leaser.KeepAlive(c.ctx, c.leaseID)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *client) Deregister(s Service) error {
+	defer c.close()
+
+	if s.Key == "" {
+		return ErrNoKey
+	}
+	if _, err := c.cli.Delete(c.ctx, s.Key, clientv3.WithIgnoreLease()); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *client) close() error {
+	if c.leaser != nil {
+		c.leaser.Close()
+	}
+	if c.watcher != nil {
+		c.watcher.Close()
+		c.wcf()
+	}
+	return nil
+}
```

Does this function need to return an error?